### ISSUE TITLE
Opting-out of cookie tracking, without comlpete opt-out of capturing

### ISSUE THREAD

**Issue by judehunter** [PostHog/posthog-js#537]
Hi, I love this tool 😄, however I couldn't find the solution to my use case in your documentation.

I'm trying to implement a cookie banner. If the user decides to opt-out of cookies (and also before the user makes any decision), I would like to set the persistence method to `"memory"`. And if they decide to opt-in, I would like to set persistence to `"cookie"`.

It looks to me like the current `.opt_in_capturing()` and `.opt_out_capturing()` methods, which are used in your React cookie banner example, enable/disable all tracking and capturing entirely. However I believe it should be possible within GDPR to simply start off with memory persistence, and in case the user opts-in to cookies, the information about the user is moved to cookies.

I'd appreciate it if you could point me to the right direction in terms of what methods to use, or if this use case isn't supported, I'd ask you to consider it.

**Comment by vetalcore** [comment]
Hey @judehunter, any luck on resolving this one?

**Comment by judehunter** [comment]
@vetalcore nope. I asked on their slack but couldn't get to a solution. I looked through the source code and as I understood it, this simoly is not supported at all, currently.

**Comment by neilkakkar** [comment]
Hey @judehunter while this is not supported out of the box, you can simulate the same:

1. Initialise posthog with 'memory' persistence
2. Have the cookie banner show up. If user opts-out, do nothing.
3. If user opts-in, then:
      1. Initialise a new PostHog client with `localStorage+cookie` persistence
      2. Disable the existing posthog client (so you don't double count things)
      3. Use the newly initialised client for all analytics

**Comment by hansottowirtz** [comment]
This makes it very hard to use this library in a "ePrivacy Directive"-compliant way. Hopefully this can fixed here or in `posthog-js-lite`.

**Comment by hansottowirtz** [comment]
It seems that when initializing with `persistence: 'cookie', disable_persistence: true,`, data is actually saved "in-memory" on `posthog.persistence.props`.

On opt in, this seems to do the trick:

```js
posthog.set_config({ disable_persistence: false });
posthog.opt_in_capturing();
posthog.persistence.save(); // writes persistence.props to a cookie
```

**Comment by neilkakkar** [comment]
Nice, that's much better!

@ivanagas this seems tutorial worthy (assuming it works for edge cases and we don't need any out-of-the-box support to make it work) :)

**Comment by feychenie** [comment]
this did not work for me, but changing it a little bit for the following worked:

```js
  posthog.set_config({ disable_persistence: false, persistence: 'localStorage+cookie' });
  posthog.opt_in_capturing();
  posthog.persistence?.save();
  ```

**Comment by ivanagas** [comment]
Hi all, I updated the [docs](https://posthog.com/docs/libraries/js#persistence) with the new `posthog.set_config` method and updated the [Next.js cookie banner](https://posthog.com/tutorials/nextjs-cookie-banner) tutorial with a full example. This enables you to opt users out of cookie tracking (`memory` mode) and then switch to `localStorage+cookies` easily.

**Comment by MichalMoravik** [comment]
Why is this compliant? We're still capturing users' IP and other personal data with `memory` mode. Are you sure the `memory` mode is fully GDPR compliant?

**Comment by ivanagas** [comment]
@MichalMoravik It's helps support compliance in terms of cookies and getting consent. We have a bunch of other privacy tools to help with compliance, including dropping users' IP if you need. More details here: https://posthog.com/docs/privacy/gdpr-compliance

**Comment by MichalMoravik** [comment]
It is presented as a GDPR compliant go-to solution and it's not. I'd expect such a nice product and great team to do much better on this front and do a proper solution for EU citizens. It should be as simple as turning a switch and start tracking and until then NO COOKIES are stored on the web. Now, even if I use "memory" sometimes I get some cookies stored from posthog.

And btw it is buggy. Sometimes it doesn't start tracking after approval, sometimes it does. It is inconsistent. I've tried different approaches, even turning off capturing by default and then turning on. Tried different options, playing around with persistence, no luck. Sadly, I liked your product so much, now it is unusable for me.
